### PR TITLE
Multi-node batched validation and improvement on strategy selection

### PR DESCRIPTION
**Pull request by hrukalive**
1. Multi-node and batched validation
    1. Binarizer changes:
        1. Split train/valid set now preserves the text prefix order specified in config.
        2. Record all ndarray's first dimension into metadata (pickle).
        3. Also print duration for each speaker separately.
        4. File handle is closed in `load_meta_data`.
    2. Task code changes: 
        1. All devices participate in validation. And diffusion process accepts a batch of inputs.
        2. Custom DsTensorBoardLogger to support logging audio and images from multiple processes.
    3. Plot changes: Added `figsize` for curves and pitch plots. Added `title` ("spk - item name")
3. Multi-device strategy selection: Much easier, does not poke PL internal logic but delegates to its registry. Using `name`, one could specify all available strategies, and pass other key-value pairs as kwargs. To disable NCCL P2P, now use a new option called `nccl_p2p` (defaults to `true`).
4. Changes due to side-effects and future considerations:
    1. Dataset now admits `size_key` to specify other "sizes" from metadata for sorting samples.
    2. Dataset now admits `preload` to load all samples into memory (for TPU).
    3. DsBatchSampler is now unified as the only sampler. Several bug fixes on unwanted randomness, and correctly deals with an insufficient number of samples.
    4. Dataset get_item and collator return the indices of the item for metadata retrieval.
    5. Moved fine-tune related logic to build model.
    6. Removed dynamic validation loss creation in validation step, hoisted to the __init__ function and conformed validation metrics to the same place.

---

## Notes

1. Even though DeepSpeed is now supported, the checkpoint produced is not in the usual format. Please use it at your own discretion.
2. FSDP does not support gradient clipping by norm.
3. This PR does require a re-binarization of the datasets.

### CODE REVIEW

File: configs/base.yaml [hunk 1]
```diff
@@ -88,8 +88,11 @@ pl_trainer_accelerator: 'auto'
 pl_trainer_devices: 'auto'
 pl_trainer_precision: '32-true'
 pl_trainer_num_nodes: 1
-pl_trainer_strategy: 'auto'
-ddp_backend: 'nccl' # choose from 'gloo', 'nccl', 'nccl_no_p2p'
+pl_trainer_strategy: 
+  name: auto
+  process_group_backend: nccl
```
Comment (yqzhishen) [inline configs/base.yaml:93]: Add these two keys in the documentation

File: docs/ConfigurationSchemas.md [hunk 2]
```diff
@@ -648,19 +648,15 @@ training
 
 #### customizability
 
-normal
+not recommended
 
 #### type
 
-str
+string
```
Comment (yqzhishen) [inline docs/ConfigurationSchemas.md:655]: str

File: preprocessing/acoustic_binarizer.py [hunk 3]
```diff
@@ -56,19 +56,19 @@ def __init__(self):
...
-                meta_data_dict[f'{ds_id}:{item_name}'] = temp_dict
+            with open(raw_data_dir / 'transcriptions.csv', 'r', encoding='utf-8') as f:
+                for utterance_label in csv.DictReader(f):
+                    item_name = utterance_label['name']
+                    temp_dict = {
+                        'wav_fn': str(raw_data_dir / 'wavs' / f'{item_name}.wav'),
+                        'ph_seq': utterance_label['ph_seq'].split(),
+                        'ph_dur': [float(x) for x in utterance_label['ph_dur'].split()],
+                        'spk_id': spk_id,
+                        'spk': self.speakers[ds_id],
```
Comment (yqzhishen) [inline preprocessing/acoustic_binarizer.py:67]: spk_name may be more suitable
